File: typescript/src/Bank.ts
import { Currency } from './Currency';
import { MissingExchangeRateError } from './MissingExchangeRateError';
import Money from "@xtrem-craft/Money";

export class Bank {
  private readonly _exchangeRates: Map<string, number> = new Map();

  /**
   * Convert from one currency to another using the provided exchange rate.
   * @param from
   * @param to
   * @param rate
   */
  static createWithExchangeRate(from: Currency, to: Currency, rate: number): Bank {
    const bank = new Bank();
    bank.addExchangeRate(from, to, rate);
    return bank;
  }

  /**
   * Add exchange rate to the bank.
   * @param from
   * @param to
   * @param rate
   */
  addExchangeRate(from: Currency, to: Currency, rate: number): void {
    this._exchangeRates.set(this.getExchangeRateKey(from, to), rate);
  }

  /**
   * Convert from one currency to another using the exchange rate provided in the bank.
   * @param to
   * @param baseMoney
   */
  convert(to: Currency, baseMoney: Money): Money {
    const fromCurrency = baseMoney.currency;
    const exchangeRateIsMissing: boolean = fromCurrency !== to && !this._exchangeRates.has(this.getExchangeRateKey(fromCurrency, to));

    if (exchangeRateIsMissing) {
      throw new MissingExchangeRateError(fromCurrency, to);
    }

    if (to === fromCurrency) {
      return baseMoney;
    }
    const rate = (this._exchangeRates.get(this.getExchangeRateKey(fromCurrency, to)) ?? 0)
    const newAmount = baseMoney.amount * rate;
    return new Money(newAmount,to);
  }

  private getExchangeRateKey(from: Currency, to: Currency): string {
    return `${from}->${to}`;
  }
}

File: python/src/bank.py
from typing import Dict
from .currency import Currency
from .missing_exchange_rate_error import MissingExchangeRateError


class Bank:
    _exchange_rate: Dict[str, float] = {}

    def __init__(self, exchange_rate = {}) -> None:
        self._exchange_rate = exchange_rate

    @staticmethod
    def create(from_currency: Currency, to_currency: Currency, rate: float) -> "Bank":
        bank = Bank({})
        bank.addEchangeRate(from_currency, to_currency, rate)

        return bank
    
    def addEchangeRate(self, from_currency: Currency, to_currency: Currency, rate: float) -> None:
        self._exchange_rate[f'{from_currency.value}->{to_currency.value}'] = rate

    def convert(self, amount: float, from_currency: Currency, to_currency: Currency) -> float:
        exchange_rate =f'{from_currency.value}->{to_currency.value}' 

        if not (from_currency.value == to_currency.value or  exchange_rate in self._exchange_rate):
            raise MissingExchangeRateError(from_currency, to_currency)
        
        if from_currency.value == to_currency.value :
            return amount
        
        return amount * self._exchange_rate[exchange_rate]